app/config/Switches.scala

```diff
@@ -1,33 +1,40 @@
 package config
 
-import java.util.concurrent.{Executors, TimeUnit}
-import _root_.utils.Notifier
-import com.amazonaws.services.s3.AmazonS3
-import com.amazonaws.services.s3.model.{GetObjectRequest, ObjectMetadata, PutObjectRequest}
-import com.amazonaws.util.StringInputStream
-import utils.Loggable
+import _root_.utils.{Loggable, Notifier}
+import com.gu.etagcaching.ETagCache
+import com.gu.etagcaching.FreshnessPolicy.TolerateOldValueWhileRefreshing
+import com.gu.etagcaching.aws.s3.ObjectId
+import com.gu.etagcaching.aws.sdkv2.s3.S3ObjectFetching
 import play.api.libs.json.{Format, JsString, JsValue, Json}
+import software.amazon.awssdk.services.s3.S3AsyncClient
 
-import java.util.concurrent.atomic.AtomicReference
+import java.util.concurrent.{Executors, TimeUnit}
 import scala.concurrent.ExecutionContext.Implicits.global
-import scala.io.Source
+import scala.concurrent.Future
+import scala.concurrent.duration._
 
-class Switches(config: LoginConfig, s3Client: AmazonS3) extends Loggable {
+class Switches(config: LoginConfig, s3AsyncClient: S3AsyncClient) extends Loggable {
 
   type SwitchMap = Map[String, SwitchState]
-  private val atomicSwitchMap: AtomicReference[SwitchMap] = new AtomicReference[SwitchMap](Map.empty)
+  val switchMapCache: ETagCache[ObjectId, SwitchMap] = new ETagCache(
+    S3ObjectFetching.byteArraysWith(s3AsyncClient)
+      .thenParsing(Json.parse(_).as[SwitchMap])
+      .onUpdate { update => logSwitchDiff(update.oldV,update.newV) },
+    TolerateOldValueWhileRefreshing,
+    _.maximumSize(1).refreshAfterWrite(1.minute)
```

Comment:
IIUC, I can't think of a good reason to have much of caching TTL at all. Perhaps a time of ten seconds or so would be appropriate?

The e-tag requests should make requests efficient and if there is a change at any time it would be sensible to retrieve the latest.

Having a refresh still seems useful to guarantee that the file gets updated and notifications are sent. If a user goes to an emergency endpoint, the switch endpoint, or the healthcheck request, it makes sense to update the switch setting as soon as possible.

Reply:
Yeah, sounds great! Done with 61aabfeeebf6166f0ec09333e47423ce61dda08f